src/goto-symex/complexity_module.cpp

```diff
@@ -0,0 +1,205 @@
+    const std::size_t unwind = options.get_signed_int_option("unwind");
+
+    // In order of priority:
+    //  * explicit limit
+    //  * inferred limit from unwind
+    //  * best limit we can apply with very little information.
+    if(failed_children_limit > 0)
+      max_loops_complexity = failed_children_limit;
+    else if(unwind > 0)
+      max_loops_complexity = std::max(static_cast<int>(floor(unwind / 3)), 1);
+    else
+      max_loops_complexity = limit;
+  }
+}
+
+/// Takes into account nested loops and various other subtleties.
+optionalt<std::reference_wrapper<framet::active_loop_infot>>
+complexity_modulet::get_current_active_loop(call_stackt &current_call_stack)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    // As we're walking bottom-up, the first frame we find with active loops
+    // is our closest active one.
+    if(!frame_iter->active_loops.empty())
+    {
+      return std::reference_wrapper<framet::active_loop_infot>(
+        frame_iter->active_loops.back());
+    }
+  }
+
+  return {};
+}
+
+bool complexity_modulet::in_blacklisted_loop(
+  const call_stackt &current_call_stack,
+  const goto_programt::const_targett &instr)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    for(auto &loop_iter : frame_iter->active_loops)
+    {
+      for(auto &blacklisted_loop : loop_iter.blacklisted_loops)
+      {
+        if(blacklisted_loop.get().contains(instr))
+        {
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+bool complexity_modulet::are_loop_children_too_complicated(
+  call_stackt &current_call_stack)
+{
+  // Check that current loops don't violate complexity limits, if they
+  // do then add that loop to the blacklist for all currently-running loops.
+
+  // We want to make sure that if a certain execution of a loop is too complex
+  // no matter the branching, that we cancel out early. This is critical because
+  // otherwise we will continue to execute nested loops that we already know
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+
+  // If non-empty, is the lowest-level loop we should blacklist in the stack.
+  optionalt<std::reference_wrapper<lexical_loopst::loopt>> loop_to_blacklist;
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    for(auto it = frame_iter->active_loops.rbegin();
+        it != frame_iter->active_loops.rend();
+        it++)
+    {
+      auto &loop_info = *it;
+
+      // If we've already violated a complexity limit, add to all parent loops
+      // our currently blacklisted ones
+      if(loop_to_blacklist)
+      {
+        loop_info.blacklisted_loops.emplace_back(*loop_to_blacklist);
```

Comment:
I don't think this works: it adds `loop_to_blacklist`, the originally blacklisted loop, again and again, rather than actually adding parents.

Reply:
Yeah, that's what it's meant to do. Because this is reversed we'll be walking bottom-up, at the point something becomes blacklisted we then add that to every parent that's still executing. Then those parents will make sure they aren't executing any blacklisted loops.